### PR TITLE
Detect breaking changes to column names and data types in state:modified check

### LINKED ISSUE #6869: [CT-2038] Detect breaking changes to column names and data types in state:modified check
If a model with `contract:true` has its column name/data types modified in a way that represents a breaking change, dbt should raise an error as part of the `state:modified` check. 

What constitutes a breaking change?
- Removal of column
- name change of column
- data_type change of column 
- changing `contract:true` to `contract:false`

### PR DESCRIPTION
**Pull request by gshank**
resolves #6869

### Description

Add "same_contract", "state:modified.contract" to defer_state checks, create new exception and tests.

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have [opened an issue to add/update docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose), or docs changes are not required/relevant for this PR
- [x] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

**Comment by gshank** [comment]
Please look at the exception message to see what you want.

File: core/dbt/exceptions.py
Comment (dbeatty10) [inline core/dbt/exceptions.py:226]: This error message looks good to me 👍

Comment (MichelleArk) [reply core/dbt/exceptions.py:226]: If the user is aware they're making a breaking change, and just needs some guidance regarding making a new version of the model - this error message looks good to me too. 

That said, I am worried that this error message could be a little opaque for unintented breaking changes, especially as [our set of conditions that constitute breaking changes becomes more nuanced](https://github.com/dbt-labs/dbt-core/issues/7065). For example, a user updates a contracted model with constraints from `table` to `view` in an attempt to optimize some spend, which raises a breaking change error because the existing constraints can't be validated. Just seeing `There is a breaking change in the model contract` could be confusing, and perhaps the user would actually choose to just revert that change and make it at a later point (bundle it with some other breaking changes). In that scenario, I'd (as an imaginary user) love to see an error message that looks more like: 
```
There is a breaking change in the model contract: 
 * Updated materialization from `table` to `view` for model with constraints
 * Updated config from contract: true to contract: false
 * Column a removed
 * Column c updated data_type from string to int
 * ...
You may need to create a new version. See: https://docs.getdbt.com/docs/collaborate/publish/model-versions
```

This may not be concern with the current set of conditions for breaking changes (any change to column name/data_type => breaking change), but given that we're planning to extend this to be more nuanced soon in https://github.com/dbt-labs/dbt-core/issues/7065, I'd leave some design space for it if this scenario seems important to prioritize the UX of (cc @jtcohen6).

Comment (gshank) [reply core/dbt/exceptions.py:226]: Currently the change of a materialization is not a contract change, it's a config change. The ticket doesn't mention materialization -- do you want that to be part of the contract change too?

Creating that more user-friendly error message will probably be about twice as much work than everything else so far. If you want me to spend my time doing that, I can.

Comment (MichelleArk) [reply core/dbt/exceptions.py:226]: Materializations are not part of this issue, but they will be part the work to extend detecting breaking changes in https://github.com/dbt-labs/dbt-core/issues/7065 (which i still blocked by https://github.com/dbt-labs/dbt-core/issues/7067). If we know we'll want more granular error messaging as part of that work, we could set up the foundations here or leave it to https://github.com/dbt-labs/dbt-core/issues/7065.

Comment (gshank) [reply core/dbt/exceptions.py]: I added a brief "reason" to the exception message. We can do more elaborate messaging once the other parts are in place.

Comment (jtcohen6) [reply core/dbt/exceptions.py]: (good conversation, no notes!)

File: core/dbt/exceptions.py
Comment (MichelleArk) [inline core/dbt/exceptions.py:226]: missing the f-string for formatting `self.reason` here.

File: tests/functional/defer_state/test_modified_state.py
Comment (MichelleArk) [inline tests/functional/defer_state/test_modified_state.py:292]: Adding a constraint shouldn't be a breaking change - just a regular change to the model contract (returns `False` from `same_contract`, but does not raise `ModelContractError`.

File: core/dbt/contracts/graph/nodes.py
Comment (MichelleArk) [inline core/dbt/contracts/graph/nodes.py:536]: Should this be a `!=` instead? I'd expect the `contract_checksum` values to be different between `self` and `old` if the model contract has changed.

File: core/dbt/contracts/graph/nodes.py
Comment (MichelleArk) [inline core/dbt/contracts/graph/nodes.py:538]: The reasons for a breaking change can occur at the same time in a state modification check. For example, column definitions changing as well as a contract going from enabled to disabled. Given that these checks are typically used in CI - could we batch up the reasons and raise a single `ModelContractError` with an array of `reasons`? That way the user would have the full picture in their CI logs.

File: core/dbt/parser/base.py
Comment (MichelleArk) [inline core/dbt/parser/base.py:317]: This should be called once the `parsed_node` has `columns`, since the contract checksum is derived from the columns. Otherwise `parsed_node.columns` will always be `{}` and produce the same checksum. I think that's what's currently happening. Perhaps we could call `build_contract_checksum` [here](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/parser/schemas.py#L940) instead, after the config and columns are patched, and constraints are validated.

File: core/dbt/contracts/graph/nodes.py
Comment (MichelleArk) [inline core/dbt/contracts/graph/nodes.py:536]: should this be `True` instead, since there is no change?

Also, it's possible to hit this line and return early even if there are elements in `breaking_change_reasons` from the first case (`contract_checksum` has not changed but `self.contract` changed from `True` to `False`). Perhaps instead of this else statement we need a separate block at the end of this method: 

```python
if breaking_change_reasons: 
   raise (ModelContractError(reasons=" and ".join(breaking_change_reasons), node=self))
else: 
  # No change
  return True
```

Comment (gshank) [reply core/dbt/contracts/graph/nodes.py:536]: Yeah, that last bit should be True. But there's another issue here, which is that right now I'm not generating the contract_checksum if contract is false, so we're not actually ever going to get both "contract has been disabled" and "column definitions have changed".  It feels kind of wrong to build the contract_checksum if contract enabled is false, but do you want it anyway?

Comment (gshank) [reply core/dbt/contracts/graph/nodes.py:536]: Well, technically it would be a generated_checksum != None. But there might not even be any column changes for that None, so it would be wrong to flag it.

Comment (gshank) [reply core/dbt/contracts/graph/nodes.py]: I've rearranged it so I think it should work now.

File: core/dbt/contracts/graph/nodes.py
Comment (MichelleArk) [inline core/dbt/contracts/graph/nodes.py:532]: Just checking my understanding: instead of building the contract checksum for every model, we only build it for models with enabled contracts or previously enabled contracts (which we detect at this point, in the state:modified check).

Comment (gshank) [reply core/dbt/contracts/graph/nodes.py:532]: That's the current state, yes. Does that feel right to you?

Comment (MichelleArk) [reply core/dbt/contracts/graph/nodes.py:532]: Yes - this feels right! Nice way to ensure we can detect the full set of breaking changes without having to unnecessarily compute a checksum for every model, even those that with unenforced contracts. The comment is definitely helpful here though! 👍

File: core/dbt/contracts/graph/nodes.py
Comment (MichelleArk) [inline core/dbt/contracts/graph/nodes.py:516]: Makes sense to leave out constraints for the scope of this issue 👍 

We should revisit whether constraints should go in the checksum when we go to implement https://github.com/dbt-labs/dbt-core/issues/7065, where:
* breaking changes to constraints should raise a `ModelContractError` from `same_contract`
* non-breaking changes to constraints (e.g. adding a constraint, removing an unenforced constraint) should return `False` from `same_contract`. 

Determining the nature of a change to a constraint will need additional logic beyond comparing checksums, and adding constraints to the checksum will make understanding changes to column data_type and names a bit trickier. I could see the checksum being a useful to determine, crudely, whether there was a _change at all_, and needing to actually inspect the columns/constraints more closely to determine the type of change - breaking vs non-breaking.

Comment (mikealfare) [reply core/dbt/contracts/graph/nodes.py:520]: What is the benefit of putting these methods on `CompiledNode` versus carrying the columns on `Contract` and overriding `__eq__()` on `Contract` (in place of `same_contract()`)?

Comment (gshank) [reply core/dbt/contracts/graph/nodes.py:520]: We are not prepared to move those fields at this time, since it would be a breaking change for packages which use those fields. In addition, not all of the attributes of a column count for breaking changes, so a simple equal doesn't work.

**Review by MichelleArk — APPROVED** [review]

**Pull request merged**
